### ISSUE TITLE
Strategy interface change discussion - enable position management and short trading

### ISSUE THREAD

**Issue by hroff-1902** [freqtrade/freqtrade#2183]
I think I'll disclose here my current ideas on design solutions for position management and trading short functionality. 

It's not a simple task, so it can be done in a steps. My main idea is the following:

0) The position management rules (algo, calculations) should be a part of the custom strategy.
1) So, as the first step, the `populate_buy()` and `populate_sell()` should be composed into one method. The resulting method (say, `populate_signals()`) should populate and return a column in the resulting dataframe, say, `position` with 2 possible values: 1 for current 'buy' and '0' for current 'sell'.
1a) The current interface should be kept. So there may be a setting like `experimental.use_populate_signal` which will switch between using current `populate_buy`/`populate_sell` and new `populate_signals`
1b) The above can probably be implemented as a helper function, which fetches results of current `populate_buy`/`populate_sell` and creates that new single signal column.
2) As the next step, this value, returned by the populate_signal, should be allowed to be float. Treated as ratio of stake_amount. So 1.0 means buy one stake_amount, 2.0 -- average up, 0.0 -- sell open position, 0.5 -- buy one half of the defines stake_amount, -1.0 -- trade short (! huh? why not...)

The user should be able to have access to trades made previously, wallets to perform position management calculations.

And voila, we are there... Of course, besides changing the strategy interface, all this will require support for fractional stakes everywhere in fraqtradebot, exchange, edge, backtesting and so on...

**Comment by xmatthias** [comment]
I think we should not mix too many things into one issue - that'll end up with an issue that noone will want to work on, since it'll be a neverending / never closing issue.

> The user should be able to have access to trades made previously, wallets to perform position management calculations

this is known, and not really related to anything else in this issue as users will benefit irrelevant of this. Tracked here, and i would like to keep that seperate: #1753
It's however dangerous to rely on since past trades are not available during backtesting.

1. i think the `position` column has flaws which may not work - what if the current candle is a "no action"? 
additionally: Using 0 as "sell" is dangerous, since we have strategies in the strategy repo "without" a sell-signal, which simply set sell to 0 (`dataframe['sell'] = 0`).
Since we have examples doing this, one has to assume that private user-strategies will do the same. 
Thinking about the logic which combines buy and sell into one column will be difficult, since 0 in the sell-column never meant "selling" - but "no action" - so you'd need to add sells as `if sell 1 then 0` - then buys `if 1 then 1` ... which complicates things beyond reasonings.
I would prefer to keep seperate buy/sell columns (see below for reasons)


2. Please don't mix in shorting here at this point (#892) - 1. is already a huge piece of work. It's good to keep it open for the future (so this should not limit adding short later - and ideally enable shorting) - but i would keep shorting tracked seperately if you don't mind.

### `position` vs. `buy`/`sell`

I would like to keep the `buy` and `sell` columns seperate for the following reasons:

* see above - problems with interfacing with old strategies

* for users, it'll be easier to clearly say "this is my sell condition" - and "this is my buy condition" without the needing to worry if the sequence is wrong (and sell overwrites buy conditions, so no buys happen). Also, signals may collide (buy and sell on the same candle) - so behaviour is different if a position is open or not (yes, this is a sign of a bad strategy...)

* combining buy/sell into `position` will limit the ability of the bot to "scale down". 
Assume have floats, with `buy=0.9` meaning we buy 90% of the configured stake. We could then allow to send a `sell=0.5` - which would sell 50% of this position (but still keep 50% of the position. This is almost impossible with only one column i think - you'd need to send `position=-0.5` - which however also means short with 50% size - which makes the code both clumsy and non-clear for users (forgot to check if position open - shorted instead of downsizing).

* buy/sell signals can still be set in one combined method (say populate signals) - but setting the columns `buy` and `sell` instead of `position`.


This became longer than i wanted it to - but i like the idea, apart from combining the buy/sell column into `position`.
TL/DR: Keep buy/sell seperate - but allow floats

**Comment by xmatthias** [comment]
Adding to the above and thinking about the future possibility to short...
maybe we should (at the same time) rename the columns from 'buy' and sell to 'entry' and 'exit'.

so `'entry'=1` is "enter long - full position" - and `'entry'=-0.8` is "enter short - 80% position"

**Comment by hroff-1902** [comment]
right, 

> -1.0 -- trade short (! huh? why not...)

**Comment by hroff-1902** [comment]
without renaming it can also be clearly explained:

"buy" means "buy position" -- i.e. open position, exit position, either long or short, both ways
"sell" means "sell position" -- i.e. liquidate position, exit position.

**Comment by hroff-1902** [comment]
actually, I propose to mimic TradingView approach, if this is feasible and possible. This would help users transfer their strategies from TradingView to freqtrade.

**Comment by xmatthias** [comment]
> without renaming it can also be clearly explained:

we can explain it as clear as we want - but `buy` will never be correct as trade-entry for a short trade - intuitively thinking, that's a sell - and the exit of a short position is a (re)buy.

We could still maintain compatibility if that's a concern - and this would be  good point to change this to "entry" ... since the whole buy/sell column interface does change.

> actually, for opening and closing the positions, I propose to mimic TradingView approach, if this is feasible and possible. This would help users transfer their strategies from TradingView to freqtrade.

then please outline this approach - i'm no pinescript developer ...

**Comment by highwaterr** [comment]
`def populate_buy_trend(self, df: DataFrame, metadata: dict) -> DataFrame:

        
        for index in range(len(df.index)):
            row = df.iloc[index]

            
            def buy():                      
                df.loc[index, "buy"] = 1

            def sell():                                               
                df.loc[index, "sell"] = 1

            if row['close'] > row['x']:
                buy()
            elif row['close'] < row['x']:
                sell()

                    
        return df`
Something like this could be used to define all signals in one function.
I'd love to see something like this being able to be used..
`def populate_signals(self, df: DataFrame, metadata: dict) -> DataFrame:

        
        for index in range(len(df.index)):
            row = df.iloc[index]

            
            def long(stake_amount):                      
                df.loc[index, "buy"] = 1
                df.loc[index, "stake_amount"] = stake_amount

            def short(stake_amount):                                               
                df.loc[index, "sell"] = 1
                df.loc[index, "stake_amount"] = stake_amount

            if row['close'] > row['x']:
                long(100)
            elif row['close'] < row['x']:
                short(-100)

                    
        return df`
The stake amount should be considered the desired end result of the trade, because shorts cancel out longs and vise versa.  So theres actually 3 states to be considered in long/short trading... long, short, and no position.  If i had a 100 long contracts open, and bot signals for a short of 100 contracts, the bot would need to issue 200 short contracts to reach the desired in result.  If we wanted to close the long of 100 contracts a `short(0)`, would issue 100 short contracts, closing the position.  Just an idea, would love to know what y'all think.

**Comment by xmatthias** [comment]
just to make sure i understand your porposal correctly - we should use 3 columns?

* `"long"` 
* `"short"`
* `"stake_amount"`

I'm summarizing this since i think that the code itself has some problems (you loop the dataframe, which is inheritly slow) - this will run into huge performance problems once you start backtesting anything > a small timerange.

> The stake amount should be considered the desired end result of the trade, because shorts cancel out longs and vise versa. So theres actually 3 states to be considered in long/short trading... long, short, and no position.

I do think that the method in the previous posts would support this, but with just 2 columns (entry / exit) - using relative stakes.

Using absolute stakes is dangerous - since a strategy should work independently of my available capital - it should only care if i want to enter a "full" position - or half a position (i'm comfortable with). Also, avalable capital will change - if the bot does well - i'll have more to trade - (but your proposal would still use 100 ...) - or less to trade (if it is doing not so well ...).


> if i had a 100 long contracts open, and bot signals for a short of 100 contracts, the bot would need to issue 200 short contracts to reach the desired in result.

I'm not sure i understand this (i think it's a terminology question...).
Assume i trade USDT as stake.
when you talk about 100 long contracts - then you have a position of 100 USDT in the pair in question?
so by executing 100 short contracts, you have a position of 0 in that pair.

so by executing another 100 short contracts - you're at -100 USDT (?)
so 1 contract is really 1 USDT (in the above sample) (?) 


Apart from the terminology - I don't think this will work - not for backtesting.
Consider the following:
backtesting calculates all signals at once - and then iterates through the dataframes checking which signal to act on (respecting max_open_trades, roi signals, stoploss, ...).

Calculating the buy / sell column is therefore fast, since it can be a vectorized operation (executed across the full dataframe at once). If we need to calculate each signal per row in backtesting, it'll become unusably slow (the impact for hyperopt is much worse, since that's basically 100ds of backtests).

Your assumption (to sell the whole position, i need to know the position beforehand) is not supportable - since `short=1` can either mean "sell my current long position" - or "go short by 1".

**Comment by highwaterr** [comment]
You’re right the % would be better.  My request of the stake amount being the desired result, enables more usability.  Without it you can only short from a “no position” state.  Wouldn’t using entry exit run into the same issues of referencing the previous stake amount? You’d need to know if it’s long or short

**Comment by highwaterr** [comment]
And to sum up the desired result.  If I’m in a long x% of my balance. I have two options: close the long and wait for a new signal or it might be a good short opportunity.  Both scenarios require you to enter a short on exchange.  A short of the same x% will close the position, wait for next signal.  But if I want to be short from that long position i need to enter a short of (x*2)%,  which closes the long and puts me in a short trade.

**Comment by xmatthias** [comment]
> But if I want to be short from that long position i need to enter a short of (x*2)%, which closes the long and puts me in a short trade.

to me this is 2 steps - you close the long position - and go short thereafter.
If you start to do both at the same time (doesn't mean you can't do it in quick succession), tracking of p&l will become very difficult, since you may have positions where you switched from long to short and back for days without ever closing it completely ...

**Comment by highwaterr** [comment]
which ever way will work best.  But I'd like it to be an option, to be able to flip long/short as well as long/close and short/close. that was my suggestion to use the stake size to actually control this.

**Comment by 4oplay** [comment]
Futures would be awesome. (Aswell as position size) What is the next step towards such feature? This would set freqtrade as project well ahead of any other one I saw so far. We might just start with a collection of necessary tasks to assign them and keep track of process. I think there are many individuals who would like to contribute even with their limited capabilities as developer.

**Comment by archeria1** [comment]
I'm working on the shorting logic. So far only in backtest mode. I added 2 new signals, sbuy and sshort and adapted the calc_profit and calc_profit_percent function. This seems to be working fine.

`    def calc_profit_percent(self, rate: Optional[float] = None,
                            fee: Optional[float] = None,
                            order_position=None) -> float:
        from freqtrade.strategy.interface import PosType

        # Original
        if order_position == None:
            open_trade_price = self.calc_open_trade_price()
            close_trade_price = self.calc_close_trade_price(
                rate=(rate or self.close_rate),
                fee=(fee or self.fee_close)
            )
            profit_percent = (close_trade_price / open_trade_price) - 1
            return float(f"{profit_percent:.8f}")
        else:
            # Order
            if order_position == PosType.LONG:
                profit_percent = (rate / self.open_rate) - 1
            elif order_position == PosType.SHORT:
                profit_percent = (self.open_rate / rate) - 1

            return float(f"{profit_percent:.8f}")`

Now for position sizing I thought about making a variable that control how much of you balance is used. Set it to 0.1 and it uses 10$ from your 100$. I could be overwritten like this `df.loc[index, "position_size_of_balance"] = 0.2`. This logic worked fine in another project. 
 
But I also like your idea of replacing the signals with entry and exit. Using only a single column with the amount of contracts to buy ([0, 0, -100, 0, 0, 100, ...]) might work as well. Not sure how well that would work out with vectorized operations though. I prefer to iterate the dataframe since it allows complex states and dynamically modifying indicators, speed isn't really an issue.

**Comment by bshankar** [comment]
Possible addition: #4756 

Because rise and fall of pairs is highly correlated, if a strategy uses many pairs and has 2-5 `max_open_trades` (optimal value for most strategies), I feel we need a way to sort the buy signals to choose the best trades. So I propose an additional column `entry_signal_strength` in the dataframe which if available will sort buy signals.

**Comment by CryptoRahino** [comment]
I w

> I'm working on the shorting logic. So far only in backtest mode. I added 2 new signals, sbuy and sshort and adapted the calc_profit and calc_profit_percent function. This seems to be working fine.
> 
> `
> def calc_profit_percent(self, rate: Optional[float] = None,
> 
> ```
>                         fee: Optional[float] = None,
>                         order_position=None) -> float:
>     from freqtrade.strategy.interface import PosType
>     # Original
>     if order_position == None:
>         open_trade_price = self.calc_open_trade_price()
>         close_trade_price = self.calc_close_trade_price(
>             rate=(rate or self.close_rate),
>             fee=(fee or self.fee_close)
>         )
>         profit_percent = (close_trade_price / open_trade_price) - 1
>         return float(f"{profit_percent:.8f}")
>     else:
>         # Order
>         if order_position == PosType.LONG:
>             profit_percent = (rate / self.open_rate) - 1
>         elif order_position == PosType.SHORT:
>             profit_percent = (self.open_rate / rate) - 1
> 
>         return float(f"{profit_percent:.8f}")`
> ```
> 
> Now for position sizing I thought about making a variable that control how much of you balance is used. Set it to 0.1 and it uses 10$ from your 100$. I could be overwritten like this `df.loc[index, "position_size_of_balance"] = 0.2`. This logic worked fine in another project.
> 
> But I also like your idea of replacing the signals with entry and exit. Using only a single column with the amount of contracts to buy ([0, 0, -100, 0, 0, 100, ...]) might work as well. Not sure how well that would work out with vectorized operations though. I prefer to iterate the dataframe since it allows complex states and dynamically modifying indicators, speed isn't really an issue.

I wonder if you had any progress with the short trading logic?
does it has ETA for release or target version?

**Comment by xmatthias** [comment]
> does it has ETA for release or target version?

Freqtrade is an open source project ran by volunteers.
As such, we do not assign tasks - nor do we provide any ETA - or even if it will ever be implemented - as life circumstances can change, or other interrests might change.

**Comment by Axel-CH** [comment]
Hello @xmatthias I'm also hugely interested by this features. I think I can help to implement it. As it is a huge piece it should be nice to validate the new strategy class architecture. Once it's validated we can split the work on small tasks to share the workload and work on it, according to each of us time and availability. 

Bellow you will find my proposition of Strategy Class

`class LongAndShortStrategy(IStrategy):

    def populate_open_long(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        dataframe.loc[
            (
                (dataframe['adx'] > 65) &
                (dataframe['plus_di'] > 0.5)
            ),
            'open_long'] = 1

        return dataframe

    def populate_close_long(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        dataframe.loc[
            (
                (dataframe['adx'] > 70) &
                (dataframe['minus_di'] > 0.5)
            ),
            'close_long'] = 1

        return dataframe

    def populate_open_short(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        dataframe.loc[
            (
                (dataframe['adx'] > 70) &
                (dataframe['minus_di'] > 0.6)
            ),
            'open_short'] = 1
            
        return dataframe

    def populate_close_short(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        dataframe.loc[
            (
                (dataframe['adx'] > 70) &
                (dataframe['minus_di'] > 0.5)
            ),
            'close_short'] = 1

        return dataframe`

**Explanation**
The key point here is to keep things simple and clear, while preserving the performance with vectorisation.

- We are not forced to alternate between long and short
- Both Position (short and long) can be open at same time, on the same pair allowing to do Hedge trading
- Open and close signals only concern their respective side (long or short)

Let me know what you think

**Comment by xmatthias** [comment]
To be honest - i think starting with the strategy interface is the wrong approach and can (and probably will) lead to wrong decisions.
I'm not clear yet how exactly the margin/short interfaces work for the different exchanges (and i would not take Binance as example, from what I've seen, their interface is the worst / most complex of all in this respect) - so deciding on a strategy interface now is not possible.

In general, I do not like the above proposal - as it introduces 4 new methods (without making it clear what will happen with the existing ones - so i assume they'll stay - which will be highly confusing) - and does not include any explanation about leverage.

In general, i'm tempted to have the "close / flip side" option in a callback style (smiliar to `custom_sell`) - but as said, it's unclear to me how this shall work precisely.

---

In my opinion, it must be a "step by step approach" (ideally one PR per step) - which will get us there eventually. otherwise it'll be one Huge PR changing almost everything at once, which will be very painful to both work on and review.

these steps should (roughly) be:
* add new columns for "mode" (name to define?)  and "leverage" to the database
* fix/create profit calculations to take these columns into account (including very comprehensive tests, to ensure that the calculations are correct
* start implementing one of the exchange interfaces (most likely FTX or kraken - but kraken rate-limits are a bummer)
* discuss the new strategy interface to allow using the new exchange methodologies
* Implement backtesting for this 

Obviously, each of the above steps should be a separate PR (except for the first 2, maybe).

Only once the "exchange interface implementations" PR is in draft, we can discuss about potential changes to the strategy interface.

So if you think you can help implementing this - that would be great, i'll be looking forward to your PR introducing the first 2ish points ... :)

**Comment by Axel-CH** [comment]
I understand your point about the complexity of this new strategy interface It can be confusing for beginners. In my use case, the important feature is to be able to open short and long position independently. In fact, my strategy class is only used to compute the indicators and the dataframe is sent to an API which populate the signals (buy or sell). My goal is to have two separate APIs one for short signals and another for long.

I'm currently trading with binance Future contract with a customized version of freqtrade and the leverage aspect is transparent.
Exemple:
If I configure freqtrade to use a stake amount of 1000$ and if have set a leverage 20 on binance, the exchange will flawlessly lock a margin stake of 50$

If all exchanges apply the same logic, the leverage aspect can be outside of the bot scope and managed directly by users on 
their exchange app

Thank for the added visibility of those tasks : )

**Comment by archeria1** [comment]
Since I wrote my comment over 1 year ago I started writing my own bot that supports shorting. My approach is to make a virtual exchange that simulates a perpetual futures contract exchange. I made a class that implements all the necessary function like open_order, get_position, get_upnl. The logic to calculate average entry price, liquidation price, etc. is from Bybit. The formulas are well documented.
This way I can simply change exchange from "bybit_inverse_live" to "bybit_inverse_test" and run the strategy in backtest mode.

The interface looks something like this for a short only DCA RSI strategy:

def tick(df, i, exchange, variables):
   contracts = exchange.get_contracts_pct_of_balance(0.01) // use 1% of your balance 
   if df.at[i, "rsi"] > RSI_OVERBOUGHT:
     exchange.open_order(-contracts, MARKET, tp=None, sl=None)
   if df.at[i, "rsi"] < 50:
     exchange.open_order(-exchange.get_position(), MARKET, reduce_only=True)

I also kept the populate_indicator function so I can keep using my old strategies I made in freqtrade. I just calculate the signals when I calculate the indicators and use "buy", "sell", "sbuy", "ssell" in the tick function.

**Comment by samgermain** [comment]
> customized version of freqtrade

@Axel-CH  Any chance I could use this customized version?

**Comment by samgermain** [comment]
Working on it, #5157

If you would like to contribute to the implementation, message me on discord samwow#2769

**Comment by Axel-CH** [comment]
## Proposition for Strategy interface modification

The idea is to allow user to use the new short feature while not adding too much complexity
and allowing legacy "populate_buy_trend" and "populate_sell_trend" to work

### Configuration modifications

Adding New option "enable_short_mode" in config ['true', 'false']
if enable_short_mode is true in config, "populate_buy_trend", "populate_sell_trend"
and associated "buy" "sell" collumns are not used.

New option "position_mode" in config ['one_way', 'two_sided']
One way mode
if open_long and open_short at the same time, do nothing

Two sided position mode (Hedge)
Seperate long and shorts, allow the user to open order on both sides at the same time
For exemple long and short trades opened at same time, the winning position stay open and the other one close.

!Warning!
If we build the logic only with the one way mode in mind, It will be very difficult to implement the two side position mode (Hedge) after


### New Signal naming choices (get_signal response)

**option 1**
buy, sell, buy_short, sell_short

**option 2**
buy, sell, sell_short, buy_short

**option 3**
buy_long, sell_long, sell_short, buy_short

**option 4**
open_long, close_long, open_short, close_short

**option 5**
enter_long, exit_long, enter_short, exit_short

**option 6** !(warning, one way mode only)
enter, exit, side


**Option 1 comment**
allow to keep the old buy and sell naming for long position
buy_short is used to signal the opening of a short position
The main issue of this naming is that in a short trading situation you must open a buy order to close a short position
There is a confusion introduced by the missusage of the buy workaround

**Option 2 comment**
allow to keep the old buy and sell naming for long position
sell_short is used to signal the opening of a short position
The naming is coherent as we open a sell order to short
A confusion can be created if the user did'not understood well that we open a sell order to open a short position

**Option 3 comment**
It's and option 2 variation with a naming coherance between long and short

**Option 4 and General comment**
The function "get_signal" return the signal allowing the bot to open required types of orders
In the long only situation that was used, the order type and signals was the same and this naming question was not present

Now as the short option is available, there is a disconnection between the signal and the associated order type.
That's why I think the option 4 is good, because it add a layer of abstraction that clarify the action wanted
and the actual order type opened on exchange.
It can be confusing to understand that to close a short you have to send a buy order...

if option 4 is selected we will have the following strategy function_name / dataframe signal collumn name pair

['function_name', 'dataframe_collumn_name']

'populate_open_long' / 'open_long'
'populate_close_long' / 'close_long'
'populate_open_short' / 'open_short'
'populate_close_short' / 'close_short'

**Option 5 comment**
It's and option 4 variation with "enter" and "exit" instead of "open" and "close"

**Option 6 comment**
This option seem limiting because long and short position can't be managed separately
! a good long exit is not necessarily of good short entry
! custom entry price must be seperated for long and short #5102
If this option is choosed the two_sided position mode won't be possible


### Leverage

Add leverage_long and leverage_short cols in dataframe

'populate_leverage_long' / 'leverage_long'
'populate_leverage_short' / 'leverage_short'


Default value is no leverage. We have to discuss if no leverage is corresponding to a value of 0 or 1

0 -> no leverage (1$ open a position of 1$)
1 -> 100% leverage (1$ open a position of 2$)
10 -> 1000% leverage (1$ open a position of 10$)

**Comment by samgermain** [comment]
buy_short and sell_short are confusing
The word "open" is already being used to describe unfilled limit orders, so my vote is on option 5

**Comment by Axel-CH** [comment]
### Importance of long and short trades separation, case study

I prepared a small exemple showing why it's important to seperate long and short logic

**Graph**
![long_and_short_position_management_exemple_graph](https://user-images.githubusercontent.com/7593404/127144717-d798e40a-ab39-4035-9e22-29b49670eb7e.png)

**Actions details**
![long_and_short_position_management_exemple_table](https://user-images.githubusercontent.com/7593404/127144880-b054f4de-f576-4291-aba2-a0cb7b349851.png)


**In this exemple, you can see that there is a period where both positions sides are active. Emphasizing the importance of the two sided position mode.**

**Comment by xmatthias** [comment]
#### Leverage

> 'populate_leverage_long' / 'leverage_long'
> 'populate_leverage_short' / 'leverage_short'

There's no point in having leverage as "populate" methods. 
Using "populate" to go through the whole dataframe at once gives us performance benefits for indicators, as well as for buy-singals (as we then won't have additional calls / checks for every candle).

Once the trade is open however (or the entry-point is decided), there's little point for this as we'll have method calls for every candle anyway.
Also, the vectorized approach will remove the possibility to rely on historic trades (or currently other open trades) in backtesting, which is a major pain-point for many users/strategies at this moment already.

As such, we'll be using a `leverage()` callback (naming TBD) which will provide a list of possible leverages for the pair (obviously also the side - "long" or "short") as not all pairs support all leverages - and exchanges handle this quite differently).
Having seperate callbacks might make it less complicated for some cases, where long and short logic has nothing to do with each other - in practice however, people will probably be using the same logic in 90% of the cases, and only rarely take different leverage options depending on long and short - and as such, having seperate callbacks will result in either a lot of code duplication in the strategy (same code in long and short) - or in the usage of helper methods.

---

#### column / method namings

`open` is definitely a bad word - i'd use enter / exit respectively.
I would however not create more / additional `populate_*` methods. The sepatation between `populate_buy_trend()` and `populate_sell_trend()` is only for "understanding" purposes - and in theory, you could simply return the dataframe from one method, and assign both required columns in the other (or even in `popualate_indicators()` - even though this will have side-effects for hyperopt).


I agree with not using point 6 - we'll have to have seperate columns (but currently i don't see the need for separate methods).
As such - i'd (for now) stick with `populate_sell_trend` and `populate_buy_trend` (the name of these is currently not important - they could even be consolidated into `populate_signals` - at least assuming the "old" hyperopt interface is removed at some point ... ) - and focus on the columns instead.

I'm not sure we'll need 4 columns though - we could have a "enter" column - with "1 = buy / long" - "-1 = sell / short" ... and a similar behaviour for exit.
I'm sure you'll be able to find down-sides to this ... the most likely is the fact that this will prevent "optional" hedge mode - with both sides having a position open (_added this sentence after reading the remainder of your comment_) 

Option 2 MIGHT be an option as it would be 100% compatible to the current interface - but we can "rename" a column from `buy` to `enter_long` (and `sell` -> `exit_long`) quite easily - as their meaning will remain identical (!).

> buy_short and sell_short are confusing

Couldn't agree more - what are you doing now - buying or shorting? This option is clearly a no-go (i think it's option 1 and option 3).



As such - i think we're basically only left with Option 5 - which works fine for me (although without having 4 `populate_*` methods).

---

#### Two sided mode / hedge

We should definitely design the interface in a way that _facilitates_ this (as well as position stacking / increasing / reducing a position). Neither of these (hedge mode, position modification, swap) will however be part of the initial implementation. 
We'll have to take this slow, step by step. Once "normal" short mode is stable and proven to work without major issues (and did a few 100 trades on different settings / configurations) - we can then think on looking at these options for the future - but i honestly do expect quite some problems in this initial step already, as it seems to work quite differently by exchange (or even by market - if we look at binance).


#### configuration options 

As such (see above)  - i currently don't see the need for a `"position_mode"` in the configuration (we'll only have one trade (long or short) per pair for the moment). 

I'm also not sure on `enable_short_mode` - a compatibility layer that is transparently compatible (renaming columns, eventually reassigning methods) can be introduced quite easily as long as parameters don't change drastically - and even then it can be facililtated easily.

If any - then "short-enabling" should be a protection setting (to avoid unintentional shorts) - not something that changes the behaviour of the strategy directly in the way you are proposing.
That would be a very dangerous approach - and one that'd be VERY difficult to support going forward.

**Comment by samgermain** [comment]
> having seperate callbacks will result in either a lot of code duplication in the strategy (same code in long and short) 

Wouldn't this situation have the same code in enter_long & exit_short and enter_short&exit_long ? Also I feel like 90% of bad strategies will use the same logic. If I'm doing something like BBRSI, I would want to do something like enter on a lower_band and exit on the middle_band, but the shorting logic would enter on the upper_band, and exit on the middle_band. The logic is similar, but not the same

---------------------

> As such (see above) - i currently don't see the need for a "position_mode" in the configuration (we'll only have one trade (long or short) per pair for the moment).

I don't like that idea, with custom_sell I plan to put my orders on the exchange, so that they can catch short peaks and dips(a 1m wait for the next candle is too long) and I would like to be able to set my exit trades at a price, and my enter trades at another price just past that a little bit, but set them at the same time

**Comment by xmatthias** [comment]
> having seperate callbacks will result in either a lot of code duplication in the strategy (same code in long and short)

What you really need is new columns (enter_long/enter_short/...) -  not new methods.

Please stick to 2 `populate_<>_trend()` for the moment. Spanning this out to more methods can be a possibility for later - but i think that will complicate many things without much benefits.

It's not even relevant if the logic is the same or not - in many cases, it'll simply be 2 consecutive assignments. There's no need to split that out into 2 methods - if we anyway call all methods in sequence (as we do now with `populate_buy_trend` and `populate_sell_trend`.
WHAT you assign in each of these methods is irrelevent in all instances except for the "old" hyperopt mode (with seperate files).

---

As mentioned above - having 2 orders / trades for the same pair at the same time will not be possible to begin with - so there's no reason to add a config flag that's doing nothing. If necessary, it can be added later - but as mentioned before - i expect short-mode to be experimental / fluky / faulty for months to start - so i will not complicate the code even further by mixing this in.
 
That's a seperate enhancement that is not related to shorting - and as such - will not be added as part of that.

**Comment by samgermain** [comment]
> As such, we'll be using a leverage() callback (naming TBD) which will provide a list of possible leverages for the pair (obviously also the side - "long" or "short") as not all pairs support all leverages - and exchanges handle this quite differently).

Could we just have the populate methods assign numbers other than one? And that number is the leverage?

**Comment by xmatthias** [comment]
technically - sure - but we will not be doing that.

Using a callback will allow you to get access to prior trades (also in backtesting) - which is something that no populate* method can give you.
As such - we will be using a callback (similar / identical to the custom_stake amount callback).

**Comment by samgermain** [comment]
> technically - sure - but we will not be doing that.
> 
> Using a callback will allow you to get access to prior trades (also in backtesting) - which is something that no populate* method can give you.
> As such - we will be using a callback (similar / identical to the custom_stake amount callback).

I don't really understand how to callback is going to work

**Comment by xmatthias** [comment]
look at the implementation / usage of `custom_stake_amount` ... for leverage i'd be doing it in a similar manner - calling it before opening the position / entering the trade.
